Feature/datasource ssh keys

riethm opened the pull request:
- Added a datasource `softlayer_ssh_key`, which can be used to import ssh keys into the local state, using `label` as the import key
- Removed the `ssh_key_labels` property from the softlayer_virtual_guest resource schema
- Renamed `name` to `label` in the softlayer_ssh_key resource schema, to reflect the name of the field as is commonly visible to users in the portal or CLI (`name` is an API field)

riethm committed:
Add datasource for ssh_keys

riethm committed:
Add field descriptions and logic for handling multiple matching labels

riethm committed:
Add docs

riethm committed:
Add acceptance tests

riethm committed:
Rename ssh_keys.name to ssh_keys.label

riethm committed:
Mark ssh_key_labels field as Removed

riethm committed:
Update docs

riethm committed:
Reword a couple of attribute descriptions for the ssk_key datasource

riethm committed:
Additional usage information for ssh key data source

renier commented on docs/datasources/softlayer_ssh_key.md:
```diff
@@ -0,0 +1,41 @@
+# `softlayer_ssh_key`
+
+Use this data source to import the details of an *existing* SSH key as a read-only data source.
```
I would add that you can use this to get the SSH key id that you could use in other resources like virtual_guest.

renier commented on docs/datasources/softlayer_ssh_key.md:
````diff
@@ -0,0 +1,41 @@
+```hcl
+resource "softlayer_virtual_guest" "vm1" {
+    ...
+    ssh_keys = ["${data.softlayer_ssh_key.public_key.id}"]
````
will this result in a string or an int?

riethm replied:
Good question - I have never been able to get it to work with quotes omitted.  Here's what I get when I use this (quoted) syntax:

```
ssh_keys.#:               "1"
ssh_keys.0:               "383111"
```

So clearly, string.  But `ssh_keys = [383111]` yields the exact same result.  Maybe Terraform stores everything as `interface{}`, displays as strings, and type asserts as needed.

renier replied:
ugh, and `ssh_keys = [${data.softlayer_ssh_key.public_key.id}]` would not work then? That means I made a mistake in some of the doc updates.

renier commented on docs/datasources/softlayer_ssh_key.md:
```diff
@@ -0,0 +1,41 @@
+* `most_recent` - (Optional) If more than SSH key matches the label, use the most recent key
+
+NOTE: If more or less than a single match is returned by the search, Terraform will fail.
+Ensure that your search is specific enough to return a single SSH key only,
```
Ensure that your label...

riethm replied:
Fixed

renier commented on docs/datasources/softlayer_ssh_key.md:
```diff
@@ -0,0 +1,41 @@
+
+NOTE: If more or less than a single match is returned by the search, Terraform will fail.
+Ensure that your search is specific enough to return a single SSH key only,
+or use most_recent to choose the most recent one.
```
or use `*most_resent*`...

riethm replied:
Fixed

renier commented on docs/softlayer_virtual_guest.md:
```diff
@@ -118,10 +118,9 @@ The following arguments are supported:
     * *Default*: nil
     * *Optional*
     * *Conflicts with* `ssh_key_labels`
```
Need to remove this line, too.

riethm replied:
Fixed

renier commented on softlayer/resource_softlayer_ssh_key.go:
```diff
@@ -28,11 +28,17 @@ func resourceSoftLayerSSHKey() *schema.Resource {
+				Optional: true,
+				Removed:  "Use 'label' instead",
+			},
+
```
I would not bother with deprecations... Up to you. and it's nice that they have _Removed_ for these situations, but I don't think we are there yet to need it.

riethm replied:
Removed `Removed`

renier commented on softlayer/resource_softlayer_virtual_guest.go:
```diff
@@ -187,17 +187,16 @@ func resourceSoftLayerVirtualGuest() *schema.Resource {
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Removed:  fmt.Sprintf("Use the interpolation method described at %s", datasourceSshKeyDoc),
 			},
```
This has been there for such a short time, I'd just remove.

riethm replied:
Removed `Removed`

renier reviewed (commented):
I think we need to evolve our docs directory. Looking at the terraform project, they have the provider documentation, then two subdirectories. One for resources, another for data sources. I would like to see us emulate that.

riethm committed:
Address review recommendations

Remove Deprecated fields - deemed unnecessary at this time
Minor documentation fixes

riethm committed:
Gofmt
